Rubberduck.VBEEditor/Events/VbeEvents.cs

```diff
@@ -0,0 +1,222 @@
+                components.ComponentSelected -= ComponentSelectedHandler;
+                components.ComponentReloaded -= ComponentReloadedHandler;
+                components.DetachEvents();
+
+                _components.Remove(projectId);
+            }
+        }
+
+        public event EventHandler<ProjectEventArgs> ProjectAdded;
+        private void ProjectAddedHandler(object sender, ProjectEventArgs e)
+        {
+            if (!_components.ContainsKey(e.ProjectId))
+            {
+                RegisterComponents(e.Project);
+            }
+            ProjectAdded?.Invoke(sender, e);
+        }
+
+        public event EventHandler<ProjectEventArgs> ProjectRemoved;
+        private void ProjectRemovedHandler(object sender, ProjectEventArgs e)
+        {
+            UnregisterComponents(e.ProjectId);
+            ProjectRemoved?.Invoke(sender, e);
+        }
+
+        public event EventHandler<ProjectRenamedEventArgs> ProjectRenamed; 
+        private void ProjectRenamedHandler(object sender, ProjectRenamedEventArgs e)
+        {
+            ProjectRenamed?.Invoke(sender, e);
+        }
+
+        public event EventHandler<ProjectEventArgs> ProjectActivated; 
+        private void ProjectActivatedHandler(object sender, ProjectEventArgs e)
+        {
+            ProjectActivated?.Invoke(sender, e);
+        }
+
+        public event EventHandler<ComponentEventArgs> ComponentAdded;
+        private void ComponentAddedHandler(object sender, ComponentEventArgs e)
+        {
+            ComponentAdded?.Invoke(sender, e);
+        }
+
+        public event EventHandler<ComponentEventArgs> ComponentRemoved; 
+        private void ComponentRemovedHandler(object sender, ComponentEventArgs e)
+        {
+            ComponentRemoved?.Invoke(sender, e);
+        }
+
+        public event EventHandler<ComponentRenamedEventArgs> ComponentRenamed; 
+        private void ComponentRenamedHandler(object sender, ComponentRenamedEventArgs e)
+        {
+            ComponentRenamed?.Invoke(sender, e);
+        }
+
+        public event EventHandler<ComponentEventArgs> ComponentActivated;
+        private void ComponentActivatedHandler(object sender, ComponentEventArgs e)
+        {
+            ComponentActivated?.Invoke(sender, e);
+        }
+
+        public event EventHandler<ComponentEventArgs> ComponentSelected;
+        private void ComponentSelectedHandler(object sender, ComponentEventArgs e)
+        {
+            ComponentSelected?.Invoke(sender, e);
+        }
+
+        public event EventHandler<ComponentEventArgs> ComponentReloaded; 
+        private void ComponentReloadedHandler(object sender, ComponentEventArgs e)
+        {
+            ComponentReloaded?.Invoke(sender, e);
+        }
+
+        public event EventHandler EventsTerminated;
+
+        #region IDisposable
+
+        private bool _disposed;
+
+        private void Dispose(bool disposing)
+        {
+            if (!_disposed && _projects != null)
```

Comment:
should probably acquire a lock here, no?

Reply:
No, this did start out implementing `IDisposable` but I later removed the `IDisposable`, made this method `private` so that `Terminate` (which already has a lock) is the *only* place that can dispose this singleton. So it's safe to assume there is already a lock present by time we enter this method. I suppose I might need to rename the method now that it's no longer a true disposing pattern.

Reply:
Nah, it's close enough to earn the name ;-)
I just somehow glossed over the `private` modifier